README.txt
A little addon to read data from external sources and add that information to the Kodi weather window properties.

This script is designed as run and done.  The external sensor or some other script will have to trigger it to update the weather window properties.  The script has a plugin architecture so that additional sensor types can be easily added.  Plugins must support two method calls:

GetSensorInfo()
This method must return a list of dicts.  Each dict will be parsed, and a weather window property will be set with the dict reference being the property name and the dict value being the property value.  My Estuary modification uses IndoorTemp, IndoorHumidity, IndoorPressure, and AutoDim (string values of True or False).


HandlePassback( action )
The main script can accept arguments in the form of action=thing.  HandlePassback will be sent the specific action and needs to respond to it if possible.  HandlePassback returns a string telling the main script what happened (mostly just for debug logging).  My Estuary Mod currently uses the following actions:

BrightnessUp
BrightnessDown
AutoDimOn
AutoDimOff


Special Cases
The script can set an AutoDim weather window property either based on information passed back by it's plugins or by arguments sent when the script is run.  If you want the plugin to change the AutoDim status, send AutoDim with True or False in the returned list of dicts.  For script arguments:

action=AutoDimOn (sets AutoDim window property to the string True)
action=AutoDimOff (sets AutoDim window property to the string False)



changelog.txt
v.1.4.0
- made light level for actions more granular

v.1.3.0
- added settings to pick hardware sensors from Kodi

v.1.2.0
- changed process to a monitor
- fixes for better communication with backend

v.1.1.2
- changed logging back so nothing is logged unless debug logging is on

v.1.1.1
- restructured addon to match other development work

v,1.1.0
- Python 2/3 compatibility
- updated common libraries

v.1.0.1
- updated common libraries

v.1.0.0
- handle None sent from sensors to blank that item on the display
- removed all non-debug logging
- all data communications now done via WebSockets
- renamed SenseHAT plugin to rpi-weatherstation-lite
- pressure trend calculations no longer done in rpiwl plugin
- removed unneeded settings

v.0.2.1
- updates due to name change

v.0.2.0
- added option to hide mouse cursor (skin must provide needed files)

v.0.1.1
- new handling off screen status function and messages

v.0.1.0
- feature complete
- now sends plugins passback information based on skin buttons
- can set AutoDim status based on script arguments (see README)
- external sensor or some other script must trigger this periodically

v.0.0.5
- changed default settings

v.0.0.4
- fix for SenseHAT plugin crash

v.0.0.3
- added option for external sensors to trigger addon to update the weather window

v.0.0.2
- added pressure trend calculation to SenseHAT plugin

v.0.0.1
-initial upload

default.py
# *  Credits:
# *
# *  original Weather Station Lite Code by pkscout

from kodi_six import xbmc, xbmcaddon, xbmcgui, xbmcvfs
import os, sys
from resources.common.xlogger import Logger
from resources.common.kodisettings import getSettingBool, getSettingInt, getSettingNumber, getSettingString
import resources.plugins

addon        = xbmcaddon.Addon()
addonname    = addon.getAddonInfo('id')
addonversion = addon.getAddonInfo('version')
addonpath    = addon.getAddonInfo('path')
addonicon    = xbmc.translatePath('%s/icon.png' % addonpath )
language     = addon.getLocalizedString
preamble     = '[WS Lite]'
logdebug     = getSettingBool( addon, 'logging' ) 

lw = Logger( preamble=preamble, logdebug=logdebug )

# this section imports all the weather sensor plugins and initializes them
plugins = {'names':[], 'objs':{}}
for module in resources.plugins.__all__:
    full_plugin = 'resources.plugins.' + module
    __import__( full_plugin )
    imp_plugin = sys.modules[ full_plugin ]
    lw.log( ['loaded plugin ' + module] )
    plugin = imp_plugin.objectConfig( addon=addon )
    plugins['objs'][module] = plugin
    plugins['names'].append( [1, module] )
    lw.log( ['added %s to weather sensor plugins' % module] )



class Main:

    def __init__( self ):
        self.WINDOWID = 12600
        self.WINDOW = xbmcgui.Window( int(self.WINDOWID) )
   

    def GetSensorInfo( self, target=None, data=None ):
        try:
            sensordata = plugins['objs'][target].getSensorData( data = data )
        except KeyError:
            lw.log( ['no plugin found with the name ' + target] )
            sensordata = []
        self._set_properties( sensordata )


    def HideCursor( self ):
        self._replace_pointer_xml( 'Pointer-Hidden.xml' )
        
    
    def ShowCursor( self ):
        self._replace_pointer_xml( 'Pointer-Visible.xml' )


    def Passback( self, action ):
        try:
            plugins['names'].sort( key=lambda x: x[0] )
        except TypeError:
            pass
        for plugin_name in plugins['names']:
            data = ''
            if action == 'AutoDimOn':
                self._set_property( 'AutoDim', 'True' )
            elif action == 'AutoDimOff':
                self._set_property( 'AutoDim', 'False' )
            elif action == 'ScreenOff':
                self._set_property( 'ScreenStatus', 'Off' )
            elif action == 'ScreenOn':
                self._set_property( 'ScreenStatus', 'On' )
            lw.log( ['passing action %s to %s' % (action, plugin_name[1])] )
            result = plugins['objs'][plugin_name[1]].handlePassback( action )
            lw.log( ['got %s back from %s' % (result, plugin_name[1])] )


    def _replace_pointer_xml( self, newfile ):
        addon_dir_path = os.path.join( addonpath, '..' )
        skin_xml_path = os.path.join( addon_dir_path, xbmc.getSkinDir(), 'xml' )
        current_xml = os.path.join( skin_xml_path, 'Pointer.xml' )
        replacement_xml = os.path.join( skin_xml_path, newfile )
        lw.log( ['copying %s to %s' % (replacement_xml, current_xml)] )
        xbmcvfs.copy( replacement_xml, current_xml )


    def _set_properties( self, properties ):
        for property in properties:
            self._set_property( property[0], property[1] )
        

    def _set_property( self, property_name, value=""):
        #sets a property (or clears it if no value is supplied)
        #does not crash if e.g. the window no longer exists.
        try:
          self.WINDOW.setProperty( property_name, value )
          lw.log( ['%s set to %s' % (property_name, value)] )
        except Exception, e:
          lw.log( ["Exception: Couldn't set propery " + property_name + " value " + value , e])


def _parse_argv():
    params = {}
    for arg in sys.argv:
        arg_parts = arg.split( '=' )
        try:
            params[arg_parts[0]] = arg_parts[1]
        except IndexError:
            pass
    return params.get( 'action', None), params.get( 'plugin', None), params.get( 'data', None)


if ( __name__ == "__main__" ):
    lw.log( ['script version %s started' % addonversion] )
    lw.log( ['debug logging set to %s' % logdebug] )
    ws = Main()
    action, target, data = _parse_argv()
    if action is not None:
        if action.lower() == 'updatekodi':
            ws.GetSensorInfo( target = target, data = data )
        elif action.lower() == 'hidecursor':
            ws.HideCursor()
        elif action.lower() == 'showcursor':
            ws.ShowCursor()
        else:
            ws.Passback( action )
lw.log( ['script stopped'] )


resources/lib/wslite.py

from kodi_six import xbmc, xbmcgui
import json
from resources.lib.wslsettings import loadSettings
from resources.lib.xlogger import Logger
from resources.plugins import *



class Main( xbmc.Monitor ):

    def __init__( self ):
        """Starts the background process for weather station lite."""
        xbmc.Monitor.__init__( self )
        self._init_vars()
        self.LW.log( ['background monitor version %s started' % self.SETTINGS['ADDONVERSION']], xbmc.LOGINFO )
        self.LW.log( ['debug logging set to %s' % self.SETTINGS['debug']], xbmc.LOGINFO )
        self.LW.log( ['starting external hardware interface'] )
        while not self.abortRequested():
            if self.waitForAbort( 10 ):
                break
        self.LW.log( ['background monitor version %s stopped' % self.SETTINGS['ADDONVERSION']], xbmc.LOGINFO )


    def onNotification( self, sender, method, data ):
        if 'Other.RPIWSL_VariablePass' in method:
            data = json.loads( data )
            self.LW.log( ['MONITOR METHOD: %s DATA: %s' % (str( method ), str( data ))] )
            if data == 'ScreenStatus:ScreenOff':
                self._set_property( 'ScreenStatus', 'Off' )
            else:
                self._set_properties( self.SENSORINTERFACE.getSensorData( data=data ) )
        elif 'Other.RPIWSL_SettingsRequest' in method:
            data = json.loads( data )
            self.LW.log( ['MONITOR METHOD: %s DATA: %s' % (str( method ), str( data ))] )
            self.SENSORINTERFACE.handlePassback( 'ReturningSettings', self._scrub_settings() )


    def onSettingsChanged( self ):
        self._init_vars()
        self.SENSORINTERFACE.handlePassback( 'ReturningSettings', self._scrub_settings() )


    def _init_vars( self ):
        self.WINDOWID = 12600
        self.WINDOW = xbmcgui.Window( int(self.WINDOWID) )
        self.SETTINGS = loadSettings()
        self.LW = Logger( preamble='[WS Lite Monitor]', logdebug=self.SETTINGS['debug'] )
        self.SENSORINTERFACE = self._pick_sensor_interface()


    def _pick_sensor_interface( self ):
        if self.SETTINGS['sensor_interface'] == 0:
            return rpiwsl.SensorInterface()


    def _scrub_settings( self ):
        scrubbed_settings = {}
        for thekey, thevalue in self.SETTINGS.items():
            if 'ADDON' in thekey:
                continue
            else:
                scrubbed_settings[thekey] = thevalue
        return scrubbed_settings


    def _set_properties( self, properties ):
        for oneprop in properties:
            self._set_property( oneprop[0], value=oneprop[1] )


    def _set_property( self, property_name, value='' ):
        try:
          self.WINDOW.setProperty( property_name, value )
          self.LW.log( ['%s set to %s' % (property_name, value)] )
        except Exception as e:
          self.LW.log( ['Exception: Could not set property %s to value %s' % (property_name, value), e] )


resources/plugins/rpiwsl.py
# -*- coding: utf-8 -*-

import json
from kodi_six import xbmc
from kodi_six.utils import py2_encode



class SensorInterface:

    def __init__( self ):
        """Initialize the sensor class."""
        self.TEMPSCALE = py2_encode( xbmc.getInfoLabel('System.TemperatureUnits') )


    def handlePassback( self, json_method, data ):
        if not data:
            data = {}
        message = { 'jsonrpc': '2.0',
                    'method': 'JSONRPC.NotifyAll',
                    'params': {'sender': 'Kodi', 'message': json_method, 'data': data},
                    'id': 1 }
        xbmc.executeJSONRPC( json.dumps( message ) )


    def getSensorData( self, data=None ):
        sensordata = []
        all_values = data.split(';')
        for one_value in all_values:
            s_info = one_value.split(':')
            if s_info[0].lower().endswith('temp'):
                sensordata.append( [s_info[0], self._convert_temp( s_info[1] )] )
            elif s_info[0].lower().endswith('humidity'):
                sensordata.append( [s_info[0], self._convert_humidity( s_info[1] )] )
            elif s_info[0].lower().endswith('pressure'):
                sensordata.append( [s_info[0], self._convert_pressure( s_info[1] )] )
            else:
                sensordata.append( [s_info[0], s_info[1]] )
        return sensordata


    def _convert_humidity( self, humidity ):
        if humidity.lower() == 'none':
            return ''
        return str( int( round( float( humidity ) ) ) ) + '%'


    def _convert_pressure( self, pressure ):
        if pressure.lower() == 'none':
            return ''
        if self.TEMPSCALE == '°F':
            return '%.2f' % (float( pressure ) * 0.0295301) + ' inHg'
        else:
            return pressure + ' mbar'


    def _convert_temp( self, temperature ):
        if temperature.lower() == 'none':
            return ''
        if self.TEMPSCALE == py2_encode( '°C' ):
            temp_float =  float( temperature )
        elif self.TEMPSCALE == py2_encode( '°F' ):
            temp_float = (float( temperature ) * 9/5) + 32
        elif self.TEMPSCALE == py2_encode( 'K' ):
            temp_float = float( temperature ) + 273.15
        elif self.TEMPSCALE == py2_encode( '°Ré' ):
            temp_float = float( temperature ) * 4/5
        elif self.TEMPSCALE == py2_encode( '°Ra' ):
            temp_float = float( temperature ) * 273.15 * 9/5
        elif self.TEMPSCALE == py2_encode( '°Rø' ):
            temp_float = (float( temperature ) * 21/40) + 7.5
        elif self.TEMPSCALE == py2_encode( '°De' ):
            temp_float = (100 - float( temperature )) * 3/2
        elif self.TEMPSCALE == py2_encode( '°N' ):
            temp_float = float( temperature ) * 0.33
        else:
            temp_float =  float( temperature )
        return str( int( round( temp_float ) ) )


resources/lib/wslsettings.py

from resources.lib.kodisettings import *

SETTINGSLIST = [ {'name':'auto_dim', 'default':True},
                 {'name':'fixed_brightness', 'default':'100'},
                 {'name':'dark_action', 'default':0},
                 {'name':'dark_level', 'default':'10'},
                 {'name':'dim_action', 'default':2},
                 {'name':'dim_level', 'default':'40'},
                 {'name':'bright_action', 'default':2},
                 {'name':'bright_level', 'default':'100'},
                 {'name':'sunrise_action', 'default':0},
                 {'name':'sunrise_level', 'default':'100'},
                 {'name':'sunrise_days', 'default':0},
                 {'name':'sunset_action', 'default':11},
                 {'name':'sunset_level', 'default':'50'},
                 {'name':'sunset_days', 'default':0},
                 {'name':'timer_one_action', 'default':0},
                 {'name':'timer_one_level', 'default':'100'},
                 {'name':'timer_one_time', 'default':'00:00'},
                 {'name':'timer_one_days', 'default':0},
                 {'name':'timer_two_action', 'default':0},
                 {'name':'timer_two_level', 'default':'100'},
                 {'name':'timer_two_time', 'default':'00:00'},
                 {'name':'timer_two_days', 'default':0},
                 {'name':'dark_threshold', 'default':5},
                 {'name':'bright_threshold', 'default':80},
                 {'name':'sensor_interface', 'default':0},
                 {'name':'which_camera', 'default':0},
                 {'name':'which_sensor', 'default':0},
                 {'name':'debug', 'default':False}
               ]


def loadSettings():
    settings = {}
    settings['ADDON'] = ADDON
    settings['ADDONNAME'] = ADDONNAME
    settings['ADDONLONGNAME'] = ADDONLONGNAME
    settings['ADDONVERSION'] = ADDONVERSION
    settings['ADDONPATH'] = ADDONPATH
    settings['ADDONDATAPATH'] = ADDONDATAPATH
    settings['ADDONICON'] = ADDONICON
    settings['ADDONLANGUAGE'] = ADDONLANGUAGE
    for item in SETTINGSLIST:
        if isinstance( item['default'], bool ):
            getset = getSettingBool
        elif isinstance( item['default'], int ):
            getset = getSettingInt
        elif isinstance( item['default'], float ):
            getset = getSettingNumber
        else:
            getset = getSettingString
        settings[item['name']] = getset( item['name'], item['default'] )
    return settings






resources/plugins/rpi-weatherstation-lite.py
#v.0.3.1
# -*- coding: utf-8 -*-

import json as _json
from kodi_six import xbmc
from kodi_six.utils import py2_encode, py2_decode
from resources.common.xlogger import Logger
from resources.common.kodisettings import getSettingBool, getSettingInt, getSettingNumber, getSettingString



class objectConfig():
    def __init__( self, addon ):
        self.LW = Logger( preamble='[WS Lite - Sensors]', logdebug=getSettingBool( addon, 'logging' ) )
        self.TEMPSCALE = py2_encode( xbmc.getInfoLabel('System.TemperatureUnits') )

       
    def handlePassback( self, action ):
        message = { "jsonrpc": "2.0",
                    "method": "JSONRPC.NotifyAll",
                    "params": {"sender": "Weatherstation", "message": "RPIWSL_VariablePass", "data": {"action":action}},
                    "id": 1 }
        xbmc.executeJSONRPC( _json.dumps( message ) )
        return 'passed message via websockets to rpi.weatherstation.lite'
 
    def getSensorData( self, data=None ):
        sensordata = []
        all_values = data.split(';')
        for one_value in all_values:
            s_info = one_value.split(':')
            if s_info[0].lower().endswith('temp'):
                sensordata.append( [s_info[0], self._convert_temp( s_info[1] )] )
            elif s_info[0].lower().endswith('humidity'):
                sensordata.append( [s_info[0], self._convert_humidity( s_info[1] )] )
            elif s_info[0].lower().endswith('pressure'):
                sensordata.append( [s_info[0], self._convert_pressure( s_info[1] )] )
            else:
                sensordata.append( [s_info[0], s_info[1]] )
        self.LW.log( ['returning sensor data', sensordata] )
        return sensordata

        
    def _convert_humidity( self, humidity ):
        if humidity.lower() == 'none':
            return ''
        return str( int( round( float( humidity ) ) ) ) + '%'


    def _convert_pressure( self, pressure ):
        if pressure.lower() == 'none':
            return ''
        if self.TEMPSCALE == '°F':
            return '%.2f' % (float( pressure ) * 0.0295301) + ' inHg'
        else:
            return pressure + ' mbar'
    
    
    def _convert_temp( self, temperature ):
        if temperature.lower() == 'none':
            return ''
        if self.TEMPSCALE == py2_encode( '°C' ):
            temp_float =  float( temperature )
        elif self.TEMPSCALE == py2_encode( '°F' ):
            temp_float = (float( temperature ) * 9/5) + 32
        elif self.TEMPSCALE == py2_encode( 'K' ):
            temp_float = float( temperature ) + 273.15
        elif self.TEMPSCALE == py2_encode( '°Ré' ):
            temp_float = float( temperature ) * 4/5
        elif self.TEMPSCALE == py2_encode( '°Ra' ):
            temp_float = float( temperature ) * 273.15 * 9/5
        elif self.TEMPSCALE == py2_encode( '°Rø' ):
            temp_float = (float( temperature ) * 21/40) + 7.5
        elif self.TEMPSCALE == py2_encode( '°De' ):
            temp_float = (100 - float( temperature )) * 3/2
        elif self.TEMPSCALE == py2_encode( '°N' ):
            temp_float = float( temperature ) * 0.33
        else:
            temp_float =  float( temperature )
        return str( int( round( temp_float ) ) )



resources/plugins/senseHAT.py
#v.0.2.0
# -*- coding: utf-8 -*-

import os, sys, time, xbmc
from datetime import datetime 
import _strptime # have to do to use strptime due to python bug
from ..common.fileops import readFile, checkPath
from ..common.xlogger import Logger
if sys.version_info >= (2, 7):
    import json as _json
else:
    import simplejson as _json



class objectConfig():
    def __init__( self, addon ):
        self.LW = Logger( preamble='[WS Lite - SenseHAT]', logdebug=addon.getSetting( "logging" ) )
        self.SENSORFOLDER = addon.getSetting( "folder_senseHAT" )
        self.P_RAPID = int( addon.getSetting( "rapid_pressure" ) )
        self.P_REGULAR = int( addon.getSetting( "regular_pressure" ) )
        self.P_DELTATIME = int( addon.getSetting( "p_deltatime" ) )
        self.TEMPSCALE = xbmc.getInfoLabel('System.TemperatureUnits')
        self.LOGDATEFORMAT = "%Y-%m-%d %H:%M:%S,%f"

       
    def handlePassback( self, action ):
        message = { "jsonrpc": "2.0",
                    "method": "JSONRPC.NotifyAll",
                    "params": {"sender": "Weatherstation", "message": "RPIWSL_VariablePass", "data": {"action":action}},
                    "id": 1 }
        xbmc.executeJSONRPC( _json.dumps( message ) )
        return 'passed message via websockets to rpi.weatherstation.lite'
 
    def getSensorData( self ):
        loglines = []
        sensordata = []
        f_loglines, alldata = readFile( os.path.join( self.SENSORFOLDER, 'sensordata.log' ) )
        self.LW.log( f_loglines )
        data_array = alldata.splitlines()
        try:
            last = data_array[-1]
        except IndexError:
            last = ''
        if last:
            all_values = last.split('\t')
            all_values.pop(0) #removes the unneeded date field from the sensor file
            for one_value in all_values:
                s_info = one_value.split(':')
                if s_info[0].endswith('Temp'):
                    sensordata.append( [s_info[0], self._convert_temp( s_info[1] )] )
                elif s_info[0].endswith('Humidity'):
                    sensordata.append( [s_info[0], s_info[1] + '%'] )
                elif s_info[0].endswith('Pressure'):
                    sensordata.append( [s_info[0], self._convert_pressure( s_info[1] )] )
                else:
                    sensordata.append( [s_info[0], s_info[1]] )
            pressure_trend = self._get_pressure_trend( data_array )
            if pressure_trend:
                sensordata.append( ['PressureTrend', pressure_trend] )
            self.LW.log( ['returning sensor data'] )
            self.LW.log( sensordata )
        return sensordata

        
    def _convert_pressure( self, pressure ):
        if self.TEMPSCALE == '°F':
            return '%.2f' % (float( pressure ) * 0.0295301) + ' inHg'
        else:
            return pressure + ' mbar'
    
    
    def _convert_temp( self, temperature ):
        if self.TEMPSCALE == '°C':
            temp_float =  float( temperature )
        elif self.TEMPSCALE == '°F':
            temp_float = (float( temperature ) * 9/5) + 32
        elif self.TEMPSCALE == 'K':
            temp_float = float( temperature ) + 273.15
        elif self.TEMPSCALE == '°Ré':
            temp_float = float( temperature ) * 4/5
        elif self.TEMPSCALE == '°Ra':
            temp_float = float( temperature ) * 273.15 * 9/5
        elif self.TEMPSCALE == '°Rø':
            temp_float = (float( temperature ) * 21/40) + 7.5
        elif self.TEMPSCALE == '°De':
            temp_float = (100 - float( temperature )) * 3/2
        elif self.TEMPSCALE == '°N':
            temp_float = float( temperature ) * 0.33
        else:
            temp_float =  float( temperature )
        return str( int( round( temp_float ) ) )


    def _get_pressure_compare_lines( self, data_array ):
        try:
            last = data_array[-1]
            n_last = data_array[-2]
        except IndexError:
            last = ''
            n_last = ''
        if not (last and n_last):
            return '', ''
        last_values = last.split('\t')
        n_last_values = n_last.split('\t')
        # doing the date conversions this way due to bug in the version of python in Kodi
        # the shorter method worked fine in an external script
        try:
            last_time = datetime.strptime(last_values[0].rstrip(), self.LOGDATEFORMAT)
        except TypeError:
            last_time = datetime(*(time.strptime(last_values[0].rstrip(), self.LOGDATEFORMAT)[0:6]))
        try:
            n_last_time = datetime.strptime(n_last_values[0].rstrip(), self.LOGDATEFORMAT)
        except TypeError:
            n_last_time = datetime(*(time.strptime(n_last_values[0].rstrip(), self.LOGDATEFORMAT)[0:6]))
        tdelta = last_time - n_last_time
        target_seconds = self.P_DELTATIME * 60
        try:
            target = data_array[-int( target_seconds/tdelta.total_seconds() )]
        except (IndexError, ZeroDivisionError) as error:
            target = data_array[0]
        return last, target        


    def _get_pressure_trend( self, data_array ):
        current, previous = self._get_pressure_compare_lines( data_array )
        self.LW.log( ['current is ' + current] )
        self.LW.log( ['previous is ' + previous] )
        try:
            current_pressure = int( current.split('\t')[3].split(':')[1] )
            previous_pressure = int( previous.split('\t')[3].split(':')[1] )
        except IndexError:
            current_pressure = 0
            previous_pressure = 0
            return ''
        diff = current_pressure - previous_pressure
        if diff < 0:
            direction = 'falling'
        else:
            direction = 'rising'
        if abs( diff ) >= self.P_RAPID:
            return 'rapidly ' + direction
        elif abs( diff ) >= self.P_REGULAR:
            return direction
        return 'steady'
